tests/unit/core/testWithUIState.js

```diff
@@ -0,0 +1,160 @@
+/* eslint-disable react/no-multi-comp, react/prop-types */
+import * as React from 'react';
+import { compose } from 'redux';
+
+import withUIState, { generateId } from 'core/withUIState';
+import { setUIState } from 'core/reducers/uiState';
+import { dispatchClientMetadata } from 'tests/unit/amo/helpers';
+import { applyUIStateChanges, shallowUntilTarget } from 'tests/unit/helpers';
+
+describe(__filename, () => {
+  describe('withUIState', () => {
+    let store;
+
+    class OverlayBase extends React.Component {
+      closeOverlay = () => {
+        this.props.setUIState({ isOpen: false });
+      };
+
+      render() {
+        const { uiState } = this.props;
+
+        if (!uiState.isOpen) {
+          return null;
+        }
+
+        return (
+          <div className="overlay">
+            <button className="close-button" onClick={this.closeOverlay}>
+              Close overlay
+            </button>
+          </div>
+        );
+      }
+    }
+
+    const Overlay = compose(
+      withUIState({
+        fileName: __filename,
+        extractId: (props) => props.id,
+        initialState: { isOpen: true },
+      }),
+    )(OverlayBase);
+
+    const render = ({ id = 'some-id', ...props } = {}) => {
+      const root = shallowUntilTarget(
+        <Overlay id={id} store={store} {...props} />,
+        OverlayBase,
+      );
+      // Apply initial state change from componentWillMount.
+      applyUIStateChanges({ root, store });
```

Comment:
Could it be because of `shallowUntilTarget`? Would a plain old `shallow` + `dive` work?

Reply:
Nope. I got the tests passing with this:

```js
const root = shallow(<EditableCollectionAddon ...>).dive().dive().dive().dive();
```

...but `applyUIStateChanges()` was still necessary.